network/src/protocols/wire/handshake/v1/mod.rs

```diff
@@ -110,87 +135,105 @@ impl fmt::Display for MessagingProtocolVersion {
     }
 
-    pub fn add(
-        &mut self,
-        messaging_protocol: MessagingProtocolVersion,
-        application_protocols: SupportedProtocols,
-    ) {
-        self.supported_protocols
-            .insert(messaging_protocol, application_protocols);
-    }
-
-    pub fn verify(&self, other: &HandshakeMsg) -> bool {
-        self.chain_id == other.chain_id && self.network_id == other.network_id
-    }
-
-    pub fn find_common_protocols(
+    /// This function:
+    /// 1. verifies that both HandshakeMsg are compatible and
+    /// 2. finds out the intersection of protocols that is supported
+    pub fn perform_handshake(
         &self,
         other: &HandshakeMsg,
-    ) -> Option<(MessagingProtocolVersion, SupportedProtocols)> {
-        // First, find the highest MessagingProtocolVersion supported by both nodes.
+    ) -> Result<(MessagingProtocolVersion, SupportedProtocols), HandshakeError> {
+        // verify that both peers are on the same chain
+        if self.chain_id != other.chain_id {
+            return Err(HandshakeError::InvalidChainId(other.chain_id));
+        }
+
+        // verify that both peers are on the same type of network
+        if self.network_id != other.network_id {
+            return Err(HandshakeError::InvalidNetworkId(other.network_id.clone()));
+        }
+
+        // first, find the highest MessagingProtocolVersion supported by both nodes.
         let mut inner = other.supported_protocols.iter().rev().peekable();
-        // Iterate over all supported protocol versions in decreasing order.
+
+        // iterate over all supported protocol versions in decreasing order.
         for (k_outer, _) in self.supported_protocols.iter().rev() {
             // Remove all elements from inner iterator that are larger than the current head of the
             // outer iterator.
             match inner.by_ref().find(|(k_inner, _)| *k_inner <= k_outer) {
                 None => {
-                    return None;
+                    break;
                 }
                 Some((k_inner, _)) if k_inner == k_outer => {
                     // Find all protocols supported by both nodes for the above protocol version.
                     // Both `self` and `other` shold have entry in map for `key`.
                     let protocols_self = self.supported_protocols.get(k_inner).unwrap();
                     let protocols_other = other.supported_protocols.get(k_inner).unwrap();
-                    return Some((
+                    return Ok((
                         *k_inner,
                         protocols_self.clone().intersection(protocols_other.clone()),
                     ));
                 }
                 _ => {}
             }
         }
-        None
+
+        // no intersection found
+        Err(HandshakeError::NoCommonProtocols)
+    }
+}
+
+impl fmt::Debug for HandshakeMsg {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self)
+    }
+}
```

Comment:
Note for myself, really should probably make a procedural macro for this, I've written this like 20 times.

Reply:
I think there's a SilentDebug macro somewhere

Reply:
not perfect, but could be copy/pasted: https://github.com/libra/libra/blob/cdbf57301aaee34d6b0102918a87826669e5917d/crypto/crypto-derive/src/lib.rs#L113

Reply:
oh wait, this is just to use display as debug?